torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -8820,4 +8821,108 @@ def celu_hook(module, _input, output):
             self.assertEqual(ddp_grads[0], local_model.fc.weight.grad)
             self.assertEqual(ddp_grads[1], local_model.fc.bias.grad)
 
+        def _test_hook_pickling(self, hook, hook_state):
+            learning_rate = 0.01
+            chkpt_file = tempfile.gettempdir() + "/checkpoint.pt"
+            rank = self.rank
+
+            input = torch.randn(1, 1, device=rank)
+            net = torch.nn.Linear(1, 5).to(rank)
+            ddp_model = torch.nn.parallel.DistributedDataParallel(
+                net,
+                device_ids=[rank]
+            )
+            dummy_ddp_model = torch.nn.parallel.DistributedDataParallel(
+                net,
+                device_ids=[rank]
+            )
+            optimizer = torch.optim.SGD(ddp_model.parameters(), lr=learning_rate)
+            ddp_model.register_comm_hook(hook_state, hook)
+
+            for _ in range(10):
+                optimizer.zero_grad()
+                loss = ddp_model(input).sum()
+                loss.backward()
+                optimizer.step()
+
+            state = {
+                'state_dict': ddp_model.state_dict(),
+                'comm_hook': pickle.dumps(hook),
+                'comm_hook_state': pickle.dumps(hook_state)
```

Comment:
I am curious if we can do
```suggestion
                'comm_hook': hook,
                'comm_hook_state': hook_state
```
and then `torch.save()` will internally recurse to pickle the `"comm_hook"` and `"comm_hook_state"`. In other words, I wonder if it is necessary to manually call `pickle.dumps()`.

(I mention this because this will be the expected API usage from the user, and preferably, the user does not need to manually call `pickle.dumps()`.)

Reply:
I appreciate you sharing an insight about `torch.save()/torch.load()`. I've looked into implementation and checked with unit test, and confirmed, that we don't need to explicitly call `pickle.dumps()` and `pickle.loads()`. I've updated PR accordingly.